lib/grant-types/password-grant-type.ts

```diff
@@ -0,0 +1,108 @@
+import { AbstractGrantType } from '.';
+import {
+  InvalidArgumentError,
+  InvalidGrantError,
+  InvalidRequestError,
+} from '../errors';
+import { Client, Token, User } from '../interfaces';
+import { Request } from '../request';
+import * as is from '../validator/is';
+
+export class PasswordGrantType extends AbstractGrantType {
+  constructor(options: any = {}) {
+    super(options);
+
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    if (!options.model.getUser) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getUser()`',
+      );
+    }
+
+    if (!options.model.saveToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `saveToken()`',
+      );
+    }
+  }
+
+  /**
+   * Retrieve the user from the model using a username/password combination.
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-4.3.2
+   */
+
+  async handle(request, client) {
+    if (!request) {
+      throw new InvalidArgumentError('Missing parameter: `request`');
+    }
+
+    if (!client) {
+      throw new InvalidArgumentError('Missing parameter: `client`');
+    }
+
+    const scope = this.getScope(request);
+    const user = await this.getUser(request);
+
+    return this.saveToken(user, client, scope);
+  }
+
+  /**
+   * Get user using a username/password combination.
+   */
+
+  async getUser(request: Request) {
+    if (!request.body.username) {
+      throw new InvalidRequestError('Missing parameter: `username`');
+    }
+
+    if (!request.body.password) {
+      throw new InvalidRequestError('Missing parameter: `password`');
+    }
+
+    if (!is.uchar(request.body.username)) {
+      throw new InvalidRequestError('Invalid parameter: `username`');
+    }
+
+    if (!is.uchar(request.body.password)) {
+      throw new InvalidRequestError('Invalid parameter: `password`');
+    }
+
+    const user = await this.model.getUser(
+      request.body.username,
+      request.body.password,
+    );
+    if (!user) {
+      throw new InvalidGrantError(
```

Comment:
As per [RFC 6749 - Section 5.2](https://tools.ietf.org/html/rfc6749#section-5.2), if client authentication fails an `InvalidClientError` should be thrown instead of `InvalidGrantError`

Reply:
@thomseddon can u suggest on this pls

Reply:
> As per [RFC 6749 - Section 5.2](https://tools.ietf.org/html/rfc6749#section-5.2), if client authentication fails an `InvalidClientError` should be thrown instead of `InvalidGrantError`

I think your understanding was wrong.

invalid_grant
               The provided authorization grant (e.g., authorization
               code, resource owner `credentials`) or refresh token is
               invalid, expired, revoked, does not match the redirection
               URI used in the authorization request, or was issued to
               another client.

Reply:
@peter-fan-cn after reading it again you are right

However I still think that the error code for that should be `401 Unauthorized`,  I understand the the spec is warranting `400 Bad Request` for `invalid_grant` but from a error usability perspective, the `400` response is misleading.

Can I suggest similar to IBM Connect's implementation:
https://www.ibm.com/support/knowledgecenter/en/SSFS6T/com.ibm.apic.toolkit.doc/ref_apionprem_oauth_codes.html#reference_hhm_4c2_j5__token_p_pass 
> *If Password or User_Name is incorrect, the following response is returned:*
>
> `401 Unauthorized { "error":"invalid_grant" }`

Reply:
that will be a breaking change to the old users.